FMB/SortInference.cpp

```diff
@@ -252,9 +253,30 @@ void SortInference::doInference()
    Array<Stack<unsigned>> varPositions(c->varCnt());
    ZIArray<unsigned> varsWithPosEq(c->varCnt());
    IntUnionFind localUF(c->varCnt()+1); // +1 to avoid it being 0.. last pos will not be used
+
+   // This set will be used to collect the variables that occur as part of an equality in every literal (implicitly becomes empty if there's a non-equality literal).
+   // We will use this information to detect maximum model sizes as any claues which is just {X = t_i} means that the sort of X is at most i
+   // We start by assuming that all variables are always in all equalities and then remove variables that are not
+   Set<unsigned> vars_always_in_equalities;
+   vars_always_in_equalities.insertFromIterator(c->getVariableIterator());
+
    for(unsigned i=0;i<c->length();i++){
      Literal* l = (*c)[i];
      if(l->isEquality()){
+
+       // Positive equality means we might  be in a { X = t_i } scenario
+       if(l->polarity()){
+         // For each variable in vars_always_in_equalities remove it if it isn't
+         Set<unsigned>::Iterator vaie_iter(vars_always_in_equalities);
+         while(vaie_iter.hasNext()){ 
+           unsigned var = vaie_iter.next();
+           if((l->nthArgument(0)->isVar() && l->nthArgument(0)->var() != var) ||
+              (l->nthArgument(1)->isVar() && l->nthArgument(1)->var() != var)){
```

Comment:
I think the code will still fail to work for a clause like:
`X = Y | X = c`
As one pass of the loop here will kick out X from vars_always_in_equalities and another will remove Y.

Please let me know if I missed something!

Reply:
Yes, it should be && instead of || here, right? The first literal shouldn't kick anything out but then the second should kick out Y.

Refactoring to use two variables should make this more logical. I'll refactor and ask you to look again.

Reply:
There is probably many sensible ways of doing this ... 

The one I had in mind would probably check for the exact shape of the current literal: 1) positive two_var_eq, 2) positive one_var_eq_non_var, 3) rest and for the exact state: a) just starting, b) "these two vars possible candidates", c) "this single var possible"; and act accordingly.